File: DoublyLinkedList.js
class DoublyLinkedListNode {
  constructor(data) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  isEmpty() {
    return this.head === null;
  }

  // Time complexity: O(1)
  insertAtHead(value) {
    if (this.head === null) {
      this.head = new DoublyLinkedListNode(value);
      this.tail = this.head;
    } else {
      const newHead = new DoublyLinkedListNode(value);
      this.head.prev = newHead;
      newHead.next = this.head;
      this.head = newHead;
    }
  }

  // Time complexity: O(1)
  insertAtTail(value) {
    if (this.tail === null) {
      this.tail = new DoublyLinkedListNode(value);
      this.head = this.tail;
    } else {
      const newTail = new DoublyLinkedListNode(value);
      newTail.prev = this.tail;
      this.tail.next = newTail;
      this.tail = newTail;
    }
  }

  // Time complexity: O(1)
  deleteAtHead() {
    let toReturn = null;
    if (this.head !== null) {
      toReturn = this.head.data;
      if (this.head === this.tail) {
        this.head = null;
        this.tail = null;
      } else {
        this.head = this.head.next;
        this.head.prev = null;
      }
    }
    return toReturn;
  }

  // Time complexity: O(1)
  deleteAtTail() {
    let toReturn = null;
    if (this.tail !== null) {
      toReturn = this.tail.data;
      if (this.head === this.tail) {
        this.head = null;
        this.tail = null;
      } else {
        this.tail = this.tail.prev;
        this.tail.next = null;
      }
    }
    return toReturn;
  }

  // Time complexity: O(n)
  find(value) {
    let currentHead = this.head;
    while (currentHead !== null) {
      if (currentHead.data === value) return true;
      currentHead = currentHead.next;
    }
    return false;
  }

  traverse() {
    let path = "";
    for (let node = this.head; node !== null; node = node.next) {
      path += node.data + " ⇆ ";
    }
    path += "null";
    return path;
  }
}

File: DoublyLinkedList.py
from typing import Optional

class ListNode:
  def __init__(self, val: int) -> None:
    self.val: int = val
    self.prev: Optional[ListNode] = None
    self.next: Optional[ListNode] = None
  
class DoublyLinkedList:
  def __init__(self) -> None:
    self.head: Optional[ListNode] = None
    self.tail: Optional[ListNode] = None
  
  def insert_head(self, val: int):
    if not self.head: 
      self.head = ListNode(val)
      self.tail = self.head
      return
    old_head = self.head
    self.head = ListNode(val)
    self.head.next = old_head
    old_head.prev = self.head
  
  def insert_tail(self, val: int):
    if not self.head:
      self.head = ListNode(val)
      self.tail = self.head
      return
    old_tail = self.tail
    self.tail = ListNode(val)
    self.tail.prev = old_tail
    old_tail.next = self.tail
  
  def delete_head(self) -> Optional[int]:
    if not self.head: return
    old_head = self.head
    if self.head is self.tail:
      self.head = self.tail = None
    self.head = self.head.next
    self.head.prev = None
    return old_head
  
  def delete_tail(self) -> Optional[int]:
    if not self.head: return
    old_tail = self.tail
    if self.head is self.tail:
      self.head = self.tail = None
    self.tail = self.tail.prev
    self.tail.next = None
    return old_tail
  
  def contains(self, val: int) -> bool:
    cur = self.head
    while cur:
      if cur.val == val: return True
      cur = cur.next
    return False

  def print_dll(self) -> str:
    cur = self.head
    path = ''
    while cur:
      path += str(cur.val) + ' ⇆ '
      cur = cur.next
    path += 'None'
